srml/staking/src/slashing.rs

```diff
@@ -0,0 +1,772 @@
+//! where a validator releases many misbehaviors at once, or goes "back in time" to misbehave in
+//! eras that have already passed, would you encounter situations where a slashing span
+//! has multiple misbehaviors. However, accounting for such cases is necessary
+//! to deter a class of "rage-quit" attacks.
+//!
+//! Based on research at https://research.web3.foundation/en/latest/polkadot/slashing/npos/
+
+use super::{
+	EraIndex, Trait, Module, Store, BalanceOf, Exposure, Perbill, SessionInterface,
+	NegativeImbalanceOf,
+};
+use sr_primitives::traits::{Zero, Saturating};
+use support::{
+	StorageValue, StorageMap, StorageDoubleMap,
+	traits::{Currency, OnUnbalanced, Imbalance},
+};
+use rstd::vec::Vec;
+use codec::{Encode, Decode};
+
+/// The proportion of the slashing reward to be paid out on the first slashing detection.
+/// This is f_1 in the paper.
+const REWARD_F1: Perbill = Perbill::from_percent(50);
+
+/// The index of a slashing span - unique to each stash.
+pub(crate) type SpanIndex = u32;
+
+// A range of start..end eras for a slashing span.
+#[derive(Encode, Decode)]
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub(crate) struct SlashingSpan {
+	pub(crate) index: SpanIndex,
+	pub(crate) start: EraIndex,
+	pub(crate) length: Option<EraIndex>, // the ongoing slashing span has indeterminate length.
+}
+
+impl SlashingSpan {
+	fn contains_era(&self, era: EraIndex) -> bool {
+		self.start <= era && self.length.map_or(true, |l| self.start + l > era)
+	}
+}
+
+/// An encoding of all of a nominator's slashing spans.
+#[derive(Encode, Decode)]
+pub struct SlashingSpans {
+	// the index of the current slashing span of the nominator. different for
+	// every stash, resets when the account hits free balance 0.
+	span_index: SpanIndex,
+	// the start era of the most recent (ongoing) slashing span.
+	last_start: EraIndex,
+	// all prior slashing spans start indices, in reverse order (most recent first)
+	// encoded as offsets relative to the slashing span after it.
+	prior: Vec<EraIndex>,
+}
+
+impl SlashingSpans {
+	// creates a new record of slashing spans for a stash, starting at the beginning
+	// of the bonding period, relative to now.
+	fn new(window_start: EraIndex) -> Self {
+		SlashingSpans {
+			span_index: 0,
+			last_start: window_start,
+			prior: Vec::new(),
+		}
+	}
+
+	// update the slashing spans to reflect the start of a new span at the era after `now`
+	// returns `true` if a new span was started, `false` otherwise. `false` indicates
+	// that internal state is unchanged.
+	fn end_span(&mut self, now: EraIndex) -> bool {
+		let next_start = now + 1;
+		if next_start <= self.last_start { return false }
+
+		let last_length = next_start - self.last_start;
+		self.prior.insert(0, last_length);
+		self.last_start = next_start;
+		self.span_index += 1;
+		true
+	}
+
+	// an iterator over all slashing spans in _reverse_ order - most recent first.
+	pub(crate) fn iter(&'_ self) -> impl Iterator<Item = SlashingSpan> + '_ {
```

Comment:
Maybe call it `rev_iter` then?

Reply:
This data structure is canonically iterated in reverse order, so I think it's better to call it `iter`.